File: aoc1/aoc1.py
# Read text input file
list1, list2 = [], []
with open("input.txt") as file:
    for line in file:
        num1, num2 = map(int, line.split())
        list1.append(num1)
        list2.append(num2)

list1, list2 = sorted(list1), sorted(list2)

# Part 1 , distance score
distance_score = [abs(a - b) for a,b in zip(list1, list2)]

# Part 2, similarity score
similarity_score = sum([x*list2.count(x) for x in list1])

File: aoc1/aoc1.hs
import System.IO
import Data.List (sort)

readline :: String -> (Int, Int)
readline line =
    let [num1, num2] = map read (words line)
    in (num1, num2)

processFile :: FilePath -> IO ([Int], [Int])
processFile filename = do
    contents <- lines <$> readFile filename
    let tuples = map readline contents
    let (list1, list2) = unzip tuples
    pure (sort list1, sort list2)

differenceScore :: ([Int], [Int]) -> Int
differenceScore (list1, list2) =
    sum $ zipWith (\a b -> abs (a - b)) list1 list2

similarityScore :: ([Int], [Int]) -> Int
similarityScore (list1, list2) =
    sum [a * countIn a list2 | a <- list1]

countIn :: Int -> [Int] -> Int
countIn x xs = length (filter (== x) xs)

main :: IO ()
main = do
    result <- processFile "input.txt"
    print $ differenceScore result
    print $ similarityScore result
